packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,191 @@
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of deleted interval
+	// just use positions on the interval?
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: interval.start,
+		end: interval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of previous interval
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: previousInterval.start,
+		end: previousInterval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: IntervalRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	// might be better to assert if id is undefined
+	const label = revertible.interval.properties.label;
+	// should I use revertible.collection here or access from sharedstring?
+	string.getIntervalCollection(label).removeIntervalById(id);
+}
+
+function revertLocalDelete(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.DELETE>,
+) {
+	const label = revertible.interval.properties.label;
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const type = revertible.interval.intervalType;
+	const props = revertible.interval.properties;
+	string.getIntervalCollection(label).add(start, end, type, props);
+}
+
+function revertLocalChange(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.CHANGE>,
+) {
+	// TODO: remove new interval and restore previous interval
+	const label = revertible.interval.properties.label;
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	// might be better to assert if id is undefined
+	string.getIntervalCollection(label).removeIntervalById(id);
+
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const type = revertible.interval.intervalType;
+	const props = revertible.interval.properties;
+	string.getIntervalCollection(label).add(start, end, type, props);
```

Comment:
It would be better to change/move the interval rather than delete and add. This was will emit the wrong event and lost properties that were edited in the meantime.